### PR TITLE
Improve Cleanup of Jupyter processes on Notebook and/or ADS Close

### PR DESCRIPTION
**Pull request by chlafreniere**
Partial fix for #3477.

Changes include:
- When calling process.kill (if the process still exists), terminate the process non-gracefully
- When ADS closes, ensure that we call jupyter notebook stop before killing the child process (otherwise, 'jupyter notebook list' will think there's a jupyter notebook still running on a certain port after the process ends)
- Don't notify the user if there was a problem when closing the notebook (if there is a problem, seems unlikely that the user could take any action anyway)
- Always try to kill the child processes even if there was a problem when sending a message to Jupyter to stop
- Delete a handle to the existing manager when calling close notebook in ExtHostNotebook

I will investigate whether we can reduce the number of Jupyter processes in a bit.

### CODE REVIEW

**Comment by Charles-Gagnon** [comment]
> 	public checkProcessDied(childProcess: ChildProcess): void {

nit: But this isn't a great name for this method since it doesn't indicate that we kill the process if it doesn't exit gracefully in time. Something like ensureProcessEnded would be better IMO

---
Refers to: extensions/notebook/src/jupyter/serverInstance.ts:79 in 3e3c964. [](commit_id = 3e3c964052d29c7842453b28c09527d115361281, deletion_comment = False)

File: extensions/notebook/src/jupyter/serverInstance.ts [hunk 1]
```diff
@@ -156,13 +157,14 @@ export class PerNotebookServerInstance implements IServerInstance {
...
-				this._isStarted = false;
-				this.utils.checkProcessDied(this.childProcess);
-				this.handleConnectionClosed();
 			}
 		} catch (error) {
 			// For now, we don't care as this is non-critical
 			this.notify(this.options.install, localize('serverStopError', 'Error stopping Notebook Server: {0}', utils.getErrorMessage(error)));
+		} finally {
+			this._isStarted = false;
+			this.utils.checkProcessDied(this.childProcess);
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/jupyter/serverInstance.ts]: >this.utils.checkProcessDied(this.childProcess); [](start = 3, length = 47)

this.childProcess = undefined?

File: extensions/notebook/src/jupyter/serverInstance.ts [hunk 2]
```diff
@@ -156,13 +157,14 @@ export class PerNotebookServerInstance implements IServerInstance {
...
 				await this.utils.executeBufferedCommand(stopCommand, install.execOptions, install.outputChannel);
-				this._isStarted = false;
-				this.utils.checkProcessDied(this.childProcess);
-				this.handleConnectionClosed();
 			}
 		} catch (error) {
 			// For now, we don't care as this is non-critical
 			this.notify(this.options.install, localize('serverStopError', 'Error stopping Notebook Server: {0}', utils.getErrorMessage(error)));
+		} finally {
+			this._isStarted = false;
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/jupyter/serverInstance.ts]: >_isStarted [](start = 8, length = 10)

This is already done in handleConnectionClosed()

File: src/sql/workbench/parts/notebook/models/notebookModel.ts [hunk 3]
```diff
@@ -777,7 +777,7 @@ export class NotebookModel extends Disposable implements INotebookModel {
 			}
 			await this.shutdownActiveSession();
 		} catch (err) {
-			this.notifyError(localize('shutdownError', "An error occurred when closing the notebook: {0}", notebookUtils.getErrorMessage(err)));
+			console.log(localize('shutdownError', "An error occurred when closing the notebook: {0}", notebookUtils.getErrorMessage(err)));
```
Comment (Charles-Gagnon) [inline src/sql/workbench/parts/notebook/models/notebookModel.ts]: >localize [](start = 15, length = 8)

Don't usually localize console logging messages

File: src/sql/workbench/parts/notebook/models/notebookModel.ts [hunk 4]
```diff
@@ -777,7 +777,7 @@ export class NotebookModel extends Disposable implements INotebookModel {
 			}
 			await this.shutdownActiveSession();
```
Comment (YurongHe) [inline src/sql/workbench/parts/notebook/models/notebookModel.ts]: NotebookManager dispose() shutdowns all session again. they are dup. Should remove the dup?

Comment (chlafreniere) [reply src/sql/workbench/parts/notebook/models/notebookModel.ts]: I found a couple of cases where dispose() wasn't actually being called and tested the scenario where we shutdown the session multiple times; the second one was basically a no-op.

**Pull request merged**